### PR TITLE
Shell test infrastructure

### PR DESCRIPTION
**Pull request by aalexandrov**
*Description of changes:*

Provides support for writing integration tests of predictors and estimators working within workflow of a SageMaker shell, as well as a test module that uses a dummy `MeanPredictor` in order to tests the shell itself. 

I tried to split this readable commits, but the main point of the PR is the third one (i.e., the `test_shell` test module and the utility functions in the `shell.testutil` module).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/gluonts/shell/serve.py [hunk 1]
```diff
@@ -207,4 +207,73 @@ def predictor_factory(request) -> Predictor:
...
+
+
+class ServerFacade:
+    """
+    A convenience wrapper for sending requests and handling responses to
+    an inference server located at the given address.
+    """
+
+    @validated()
+    def __init__(self, base_address: str) -> None:
```
Comment (aalexandrov) [inline src/gluonts/shell/serve.py]: @jaheba: I was wondering whether there is a better type than `str` to model the `base_address` value?

**Comment by szha** [comment]
Job PR-224/2 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/2/index.html

**Comment by szha** [comment]
Job PR-224/3 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/3/index.html

**Comment by szha** [comment]
Job PR-224/4 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/4/index.html

**Comment by szha** [comment]
Job PR-224/5 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/5/index.html

File: src/gluonts/shell/serve.py [hunk 2]
```diff
@@ -207,4 +207,73 @@ def predictor_factory(request) -> Predictor:
...
+    def execution_parameters(self) -> dict:
+        response = requests.get(
+            url=self.url("execution-parameters"),
+            headers={
+                "Content-Type": "application/json",
+                "Accept": "application/json",
+            },
+        )
+
+        response_code = response.status_code
```
Comment (jaheba) [inline src/gluonts/shell/serve.py]: Not sure I like this line.

File: src/gluonts/shell/serve.py [hunk 3]
```diff
@@ -207,4 +207,73 @@ def predictor_factory(request) -> Predictor:
...
+
+    def invocations(
+        self, data_entries: Iterable[DataEntry], configuration: dict
+    ) -> List[dict]:
+        instances = list(map(serialize_data_entry, data_entries))
+        response = requests.post(
+            url=self.url("invocations"),
+            json={"instances": instances, "configuration": configuration},
+            headers={
+                "Content-Type": "application/json",
```
Comment (jaheba) [inline src/gluonts/shell/serve.py]: Why do you set these explicitly? `"Content-Type": "application/json"` is set automatically, if you use the `json` arg.

File: src/gluonts/shell/testutil.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+# First-party imports
+from gluonts.dataset.repository.datasets import materialize_dataset
+from gluonts.model.predictor import Predictor
+from gluonts.shell.sagemaker import ServeEnv, ServePaths, TrainEnv, TrainPaths
+from gluonts.shell.sagemaker.params import encode_sagemaker_parameters
+from gluonts.shell.serve import ServerFacade, Settings, make_gunicorn_app
+
+
+def random_port() -> int:
```
Comment (jaheba) [inline src/gluonts/shell/testutil.py]: I think free port is a better name.

File: src/gluonts/shell/testutil.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    """Returns a random unbound port."""
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
```
Comment (jaheba) [inline src/gluonts/shell/testutil.py]: I don't like single letter variables like `s`. Let's got with `socket_` or something like that here.

File: src/gluonts/shell/testutil.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        wrapping the spawned inference server.
+    """
+
+    gunicorn_app = make_gunicorn_app(env, forecaster_type, settings)
+    process = Process(target=gunicorn_app.run)
+    process.start()
+
+    endpoint = ServerFacade(
+        base_address="http://{address}:{port}".format(
+            address=str(settings.sagemaker_server_address),
```
Comment (jaheba) [inline src/gluonts/shell/testutil.py]: I think `str` is not needed, since it's automatically converted with `format`.

**Comment by szha** [comment]
Job PR-224/6 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/6/index.html

**Comment by szha** [comment]
Job PR-224/7 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/7/index.html

File: src/gluonts/shell/serve.py (hunk 3 above)
Comment (aalexandrov) [reply src/gluonts/shell/serve.py]: I prefer to be more explicit here for code readability, as the SageMaker contract expects the "Content-Type" and "Accept" headers to be present.

File: src/gluonts/shell/serve.py (hunk 1 above)
Comment (aalexandrov) [reply src/gluonts/shell/serve.py]: I was hoping to get something standard and similar to `IPv4Address` (with the addition of DNS addresses such as `example.com`).

`NewType` or a custom `WebAddress` class will do the trick, but I will have to introduce it within the `gluonts` namespace and force users to import and use it to annotate parameters bound to the `ServerFacade` initializer.

In the absence of a standard solution, I suggest to go with the broader `str` type.

**Comment by szha** [comment]
Job PR-224/8 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/8/index.html

File: src/gluonts/shell/serve.py [hunk 7]
```diff
@@ -207,4 +207,69 @@ def predictor_factory(request) -> Predictor:
...
+        self.base_address = base_address
+
+    def url(self, *parts) -> str:
+        return "/".join((self.base_address,) + parts)
+
+    def ping(self) -> bool:
+        try:
+            response = requests.get(
+                url=self.url("/ping"),
+                headers={
```
Comment (jaheba) [inline src/gluonts/shell/serve.py]: I think we don't need headers for `/ping`

File: src/gluonts/shell/serve.py [hunk 8]
```diff
@@ -207,4 +207,69 @@ def predictor_factory(request) -> Predictor:
...
+            )
+            return response.status_code == 200
+        except requests.exceptions.ConnectionError:
+            return False
+
+    def execution_parameters(self) -> dict:
+        response = requests.get(
+            url=self.url("/execution-parameters"),
+            headers={
+                "Content-Type": "application/json",
```
Comment (jaheba) [inline src/gluonts/shell/serve.py]: We don't need `Content-Type` here.

**Comment by szha** [comment]
Job PR-224/10 is complete. 
Docs are uploaded to http://gluon-ts-staging.s3-accelerate.dualstack.amazonaws.com/PR-224/10/index.html

**Pull request merged**